staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go

```diff
@@ -0,0 +1,257 @@
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if ValidateParameters(test.deckSize, test.handSize) != test.validated {
+				t.Errorf("test case %s fails", test.name)
+				return
+			}
+		})
+	}
+}
+
+func BenchmarkValidateParameters(b *testing.B) {
+	deckSize, handSize := int32(512), int32(8)
+	for i := 0; i < b.N; i++ {
+		_ = ValidateParameters(deckSize, handSize)
+	}
+}
+
+func TestShuffleAndDealWithValidation(t *testing.T) {
+	tests := []struct {
+		name      string
+		deckSize  int32
+		handSize  int32
+		pick      func(int32)
+		validated bool
+	}{
+		{
+			"deckSize is < 0",
+			-100,
+			8,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is < 0",
+			128,
+			-100,
+			func(int32) {},
+			false,
+		},
+		{
+			"deckSize is 0",
+			0,
+			8,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is 0",
+			128,
+			0,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is greater than deckSize",
+			128,
+			129,
+			func(int32) {},
+			false,
+		},
+		{
+			"deckSize: 128 handSize: 6",
+			128,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 1024 handSize: 6",
+			1024,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 512 handSize: 8",
+			512,
+			8,
+			func(int32) {},
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if (ShuffleAndDealWithValidation(rand.Uint64(), test.deckSize, test.handSize, test.pick) == nil) != test.validated {
+				t.Errorf("test case %s fails", test.name)
+				return
+			}
+		})
+	}
+}
+
+func BenchmarkShuffleAndDeal(b *testing.B) {
+	hashValue := rand.Uint64()
+	deckSize, handSize := int32(512), int32(8)
+	pick := func(int32) {}
+	for i := 0; i < b.N; i++ {
+		ShuffleAndDeal(hashValue, deckSize, handSize, pick)
+	}
+}
+
+func TestShuffleAndDealToSliceWithValidation(t *testing.T) {
+	tests := []struct {
+		name      string
+		deckSize  int32
+		handSize  int32
+		validated bool
+	}{
+		{
+			"validation fails",
+			-100,
+			-100,
+			false,
+		},
+		{
+			"deckSize == handSize == 4",
+			4,
+			4,
+			true,
+		},
+		{
+			"deckSize == handSize == 8",
+			8,
+			8,
+			true,
+		},
+		{
+			"deckSize == handSize == 10",
+			10,
+			10,
+			true,
+		},
+		{
+			"deckSize == handSize == 12",
+			12,
+			12,
+			true,
+		},
+	}
+	for _, test := range tests {
+		hashValue := rand.Uint64()
+		t.Run(test.name, func(t *testing.T) {
+			cards, err := ShuffleAndDealToSliceWithValidation(hashValue, test.deckSize, test.handSize)
+			if (err == nil) != test.validated {
+				t.Errorf("test case %s fails in validation check", test.name)
+				return
+			}
+
+			if test.validated {
+				// check cards number
+				if len(cards) != int(test.handSize) {
+					t.Errorf("test case %s fails in cards number", test.name)
+					return
+				}
+
+				// check cards duplication
+				cardMap := make(map[int32]struct{}, test.handSize)
+				for _, cardIdx := range cards {
+					cardMap[cardIdx] = struct{}{}
```

Comment:
this line could be in a `pick` function passed to `ShuffleAndDeal`, it is not necessary to use a slice of cards here.

Reply:
if no slice here, then another slice will be in `pick` function to check uniformity(no duplication and no overflow) as below.

Reply:
We need to check not only for range and duplication, we need to check for uniform distribution of hands.  For example, if evaluating uniformity when `handSize==3` we could represent a hand by a `[3]int /*elements in increasing order*/`.  We could have a `map[[3]int/*sorted*/]int` that maps each hand to the count of the times it has been seen.  `[3]int` is comparable; `[]int` is not (see https://golang.org/ref/spec#Comparison_operators).  It costs O(len(array)) to check for duplicates in a sorted array.

Reply:
hi, @MikeSpreitzer , I wrote a testcase to test uniform distribution like,
```
func TestUniformDistribution(t *testing.T) {
	deckSize, handSize := int32(128), int32(3)
	handCoordinateMap := make(map[int]int)

	allCoordinateCount := 128 * 127 * 126 / 6

	for i := 0; i < allCoordinateCount*16; i++ {
		hands := ShuffleAndDealToSlice(rand.Uint64(), deckSize, handSize)
		sort.Slice(hands, func(i, j int) bool {
			return hands[i] < hands[j]
		})
		handCoordinate := 0
		for _, hand := range hands {
			handCoordinate = handCoordinate<<7 + int(hand)
		}
		handCoordinateMap[handCoordinate]++
	}

	t.Logf("%d", len(handCoordinateMap))
	t.Logf("%d", allCoordinateCount)

	return
}
```
I think we've got a distribution map,  but I have no idea about how to check it, coverage? variance? Do you have some suggestions? 
cc @yue9944882

Reply:
I suggest starting by making a histogram.  You have a map from hand to hand occurrence count.  Make a histogram of hand occurrence counts.  Print it out.  Ideally all the mass will be at 16 (given the current code).  Maybe fail if less than 99% of the mass is in the range 15--17?